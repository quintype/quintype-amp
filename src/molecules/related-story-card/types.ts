import { Story } from "../../types/story";
import { Config } from "../../types/config";

export interface RelatedStoryCardTypes {
  story: Story;
  config: Config;
}

export interface ImageForStoryTypes {
  metadata: ImageMetadata | null;
  s3Key: string | null;
  altText: string | null;
}

interface ImageMetadata {
  width: number;
  height: number;
}
