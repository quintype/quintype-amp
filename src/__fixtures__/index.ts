import { config, publisherConfig, ampConfig, configOpts } from "./config.fixture";
import { textStory } from "./text-story.fixture";
import { storyWithManyJsEmbeds } from "./story-with-jsembeds.fixture";
import { relatedStories } from "./related-stories";
import { seo } from "./seo.fixture";
import { allElementsStory } from "./all-element-story.fixture";
import { liveBlogStory } from "./live-blog-story.fixture";

export {
  allElementsStory,
  config,
  publisherConfig,
  ampConfig,
  textStory,
  storyWithManyJsEmbeds,
  relatedStories,
  configOpts,
  seo,
  liveBlogStory
};
