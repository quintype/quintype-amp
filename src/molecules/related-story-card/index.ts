import { RelatedStoryCard } from "./related-story-card";
export { RelatedStoryCard };
