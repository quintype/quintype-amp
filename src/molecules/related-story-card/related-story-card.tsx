import React from "react";
import styled from "styled-components";
import { Image, DateTime, Spacer } from "../../atoms";
import { RelatedStoryCardTypes } from "./types";
import { getHumanizedDateTime } from "../../utils/date-time";
import { isGumlet } from "../../helpers";

const Wrapper = styled.div`
  margin: ${(props) => `0 0 ${props.theme.spacing.l} 0`};
  border-bottom: ${(props) => `2px solid ${props.theme.color.mono1}`};
`;
const StyledAnchor = styled.a`
  text-decoration: none;
`;
const Headline = styled.h1`
  font-family: ${(props) => props.theme.font.family.primary};
  font-size: ${(props) => props.theme.font.size.l};
  color: ${(props) => props.theme.color.black};
  font-weight: bold;
`;

export const RelatedStoryCard = ({ story, config }: RelatedStoryCardTypes) => {
  const {
    headline,
    url,
    "hero-image-metadata": imgMetadata,
    "hero-image-s3-key": imgS3Key,
    "hero-image-attribution": imgAttr,
    "hero-image-caption": imgCaption,
    "last-published-at": lastPublishedAt
  } = story;
  const humanizedDate = getHumanizedDateTime({
    dateFormat: "do MMM, yyyy 'at' p",
    timeZone: "Asia/Kolkata",
    timeStamp: lastPublishedAt
  });
  const imageOpts = isGumlet(config) ? { format: "auto" } : {};
  return (
    <Wrapper>
      <StyledAnchor href={url}>
        <Image
          metadata={imgMetadata}
          slug={imgS3Key}
          alt={imgCaption || imgAttr || "image"}
          useFallbackImage={true}
          opts={imageOpts}
          srcSetOpts={imageOpts}
        />
        <Headline>{headline}</Headline>
        <DateTime formattedDate={humanizedDate} />
        <Spacer token="s" />
      </StyledAnchor>
    </Wrapper>
  );
};

/**
 * RelatedStoryCard Component uses the RelatedStoryCard atomic components internally depending on a condition orelse it uses `<amp-img>` component.
 *
 *
 * @param {Object} props Object containing parameters passed to the render prop
 * @param {String} props.story Required. A story object from related stories collection.
 *  @category Molecules
 * @component
 */
