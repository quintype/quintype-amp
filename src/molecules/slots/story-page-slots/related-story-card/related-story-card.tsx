import { withStoryAndConfig } from "../../../../context";
import { getSlotFromConfig } from "../../helpers";

const RelatedStoryCardSlotBase = ({ story, config, index, relatedStory }) => {
  const relatedStoryCardRender = getSlotFromConfig({
    config,
    slotType: "story",
    slotName: "related-story-card-slot"
  });
  return relatedStoryCardRender && relatedStoryCardRender({ story, config, index, relatedStory });
};

/**
 * related-story-card-slot is a story page slot. It's placed after every related story card.
 *
 * @category Molecules
 * @module related-story-card-slot
 * @param {Props} props
 * @param {Story} props.story
 * @param {Config} props.config
 * @param {number} props.index index of the card. For example the slot coming after the 1st card will have index 0 and so on
 * @param {Story} props.relatedStory The related story that is above this slot
 */
export const RelatedStoryCardSlot = withStoryAndConfig(RelatedStoryCardSlotBase);
