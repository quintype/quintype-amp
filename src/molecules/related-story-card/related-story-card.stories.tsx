import * as React from "react";
import { storiesOf } from "@storybook/react";
import { RelatedStoryCard } from "./related-story-card";
import { Layout } from "../../atoms";
import { relatedStories } from "../../__fixtures__/related-stories";
import { config, textStory } from "../../__fixtures__";

const sampleRelatedStory = relatedStories[0];

storiesOf("Related Story Card", module)
  .addDecorator((story) => (
    <Layout story={textStory} config={config}>
      {story()}
    </Layout>
  ))
  .add("Default", () => <RelatedStoryCard story={sampleRelatedStory} />);
