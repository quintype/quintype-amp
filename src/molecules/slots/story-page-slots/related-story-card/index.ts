import { RelatedStoryCardSlot } from "./related-story-card";
export { RelatedStoryCardSlot };
