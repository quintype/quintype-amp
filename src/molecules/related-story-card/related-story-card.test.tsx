import React from "react";
import { RelatedStoryCard } from "./related-story-card";
import { shallow } from "enzyme";
import { relatedStories } from "../../__fixtures__";

const sampleRelatedStory = relatedStories[0];

describe("RelatedStoryCard", () => {
  it("should render default", () => {
    const wrapper = shallow(<RelatedStoryCard story={sampleRelatedStory} fallbackSrc="#" aspectRatio={[16, 9]} />);
    expect(wrapper).toMatchSnapshot();
  });
});
