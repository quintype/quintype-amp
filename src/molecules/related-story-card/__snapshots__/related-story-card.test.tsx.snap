// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelatedStoryCard should render default 1`] = `
<styled.div>
  <styled.a
    href="https://www.barandbench.com/news/governor-must-bear-in-mind-that-purpose-of-ordering-trust-vote-is-not-to-displace-governments-sc-on-madhya-pradesh-political-crisis"
  >
    <Component
      alt="image"
      metadata={null}
      opts={
        Object {
          "format": "auto",
        }
      }
      slug={null}
      srcSetOpts={
        Object {
          "format": "auto",
        }
      }
      useFallbackImage={true}
    />
    <styled.h1>
      Governor must bear in mind that purpose of ordering trust vote is not to displace governments: SC on Madhya Pradesh Political Crisis
    </styled.h1>
    <Component
      formattedDate="13th Apr, 2020 at 5:37 PM"
    />
    <Spacer
      token="s"
    />
  </styled.a>
</styled.div>
`;
